doc: Add doc for Bluetooth mesh SAR Config models

mia-ko opened the pull request:
Adding rst files describing SAR Configuration
models, SAR Configuration Client and SAR Configuration Server.

akredalen commented on doc/connectivity/bluetooth/api/mesh/sar_cfg_srv.rst:
```diff
@@ -0,0 +1,105 @@
+
+   :kconfig:option:`CONFIG_BT_MESH_SAR_RX_ACK_DELAY_INC` + 1.5
+
+* SAR Acknowledgment Retransmissions Count: This value defines the total number of retranmissions of an acknowledgment message that the stack will additionally send when the size of segments in a segmented message is above the :kconfig:option:`CONFIG_BT_MESH_SAR_RX_SEG_THRESHOLD` value.
```
This configuration is missing its :kconfig:option: (see comment above about first mention)

akredalen commented on doc/connectivity/bluetooth/api/mesh/sar_cfg_srv.rst:
```diff
@@ -0,0 +1,105 @@
+
+* SAR Acknowledgment Retransmissions Count: This value defines the total number of retranmissions of an acknowledgment message that the stack will additionally send when the size of segments in a segmented message is above the :kconfig:option:`CONFIG_BT_MESH_SAR_RX_SEG_THRESHOLD` value.
+
+* SAR Receiver Segment Interval Step: This value defines the segments reception interval step used for delaying the transmission of an acknowledgment message after receiving a new segmnet.
```
Same here.  :kconfig:option: only used in formula which is inconsistent with other definitions in the file

akredalen replied:
Also:
```suggestion
* SAR Receiver Segment Interval Step: This value defines the segments reception interval step used for delaying the transmission of an acknowledgment message after receiving a new segment.
```

akredalen commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,372 @@
+If a segmented Upper Transport PDU is tagged with additional metadata, each Lower Transport PDU of the segmented Upper Transport PDU is
+tagged with the same metadata.
+
+Delivery of a segmented message is acknowledged by the lower transport layer of the receiving node.
```
```suggestion
Delivery of a segmented message is acknowledged by the lower transport layer of the receiving node,  while an unsegmented message delivery isn't acknowledged.
```

akredalen commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,372 @@
+tagged with the same metadata.
+
+Delivery of a segmented message is acknowledged by the lower transport layer of the receiving node.
+Delivery of an unsegmented message is not acknowledged.
```
```suggestion
```

akredalen reviewed (changes requested):
Nice work! Just a few comments / minor edits

Balaklaka commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,372 @@
+Segmentation and reassembly is a mechanism used by the lower transport layer.
+It provides a way of handling larger upper transport layer messages in a mesh network, with a purpose of enhancing the Bluetooth mesh throughput.
+
+The lower transport layer takes PDUs from the upper transport layer and sends them to the lower transport layer on a peer device.
```
```suggestion
The lower transport layer takes Upper Transport PDUs from the upper transport layer and sends them to the lower transport layer on a peer device.
```

Balaklaka commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,372 @@
+It provides a way of handling larger upper transport layer messages in a mesh network, with a purpose of enhancing the Bluetooth mesh throughput.
+
+The lower transport layer takes PDUs from the upper transport layer and sends them to the lower transport layer on a peer device.
+For longer packets, which do not fit into a single Lower Transport PDU, the lower transport layer performs segmentation, splitting the PDU
+into multiple Lower Transport PDUs.
```
```suggestion
For longer packets, which do not fit into a single Lower Transport PDU, the lower transport layer performs segmentation, splitting the Upper Transport PDU
into multiple segments.
```

Balaklaka commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,372 @@
+The lower transport layer takes PDUs from the upper transport layer and sends them to the lower transport layer on a peer device.
+For longer packets, which do not fit into a single Lower Transport PDU, the lower transport layer performs segmentation, splitting the PDU
+into multiple Lower Transport PDUs.
+The receiving lower transport layer on the other device reassembles the segments into a single upper transport layer PDU before passing it
```
```suggestion
The receiving lower transport layer on the other device reassembles the segments into a single Upper Transport PDU before passing it
```

Balaklaka commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,372 @@
+Transport PDU is canceled and the Lower Transport PDUs are delivered to the network layer.
+
+Segments can be sent to a unicast address or to a group or a virtual address.
+
```
Is it possible to merge the "Transmitting to a ...." into one and just explain what it unique for unicast address, as there is a lot of repeating text. It makes it less to read and maintain. Same for retransmition.

mia-ko committed:
doc: bluetooth: mesh: doc for SAR Config models

Adding rst files describing SAR Configuration
models, SAR Configuration Client and SAR Configuration
Server.

Signed-off-by: Mia Koen <mia.koen@nordicsemi.no>

kartben commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,238 @@
+Use the :kconfig:option:`CONFIG_BT_MESH_SAR_TX_SEG_INT_STEP` Kconfig option to set a value.
+Segment transmission interval is then calculated using the following formula::
+
+(:kconfig:option:`CONFIG_BT_MESH_SAR_TX_SEG_INT_STEP` + 1) * 10 ms
```
you can't use the `:kconfig:option:` role within a code block as it doesn't render

<img width="640" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/11627bd7-f151-4506-aa51-386afbe8748a">

mia-ko replied:
Of course, I forgot to think of that. Removed the :kconfig:option:``

kartben commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,238 @@
+The initial value of the SAR Discard timer is the discard timeout value indicated by the `SAR Discard Timeout`_ state.
+
+SAR Acknowledgment timer holds the time before a Segment Acknowledgment message is sent for a received segment.
+The initial value of the SAR Acknowledgment timer is calculated using the following formula::
+
+[min(SegN + 0.5 , acknowledgment delay increment) * segment reception interval]
```
You should consider using actual math formulas as they render much more nicely.

```suggestion
The initial value of the SAR Acknowledgment timer is calculated using the following formula:

.. math::

   \min(\text{SegN} + 0.5, \text{acknowledgment delay increment}) \times \text{segment reception interval}
```

Alternatively, this simplified form works too

```suggestion
The initial value of the SAR Acknowledgment timer is calculated using the following formula:

.. math::

   min(SegN + 0.5 , acknowledgment~delay~increment) * segment~reception~interval
```

Finally, it's also possible to use the `:math:` role if you want the formula in-line with the text (``blah blah :math:`my formula` blah blah``).

kartben replied:
FWIW I would recommend you use the math directive everywhere, not just in this particular instance.
Rendered doc available at https://builds.zephyrproject.io/zephyr/pr/61886/docs/connectivity/bluetooth/api/mesh/sar_cfg.html , in case you are not aware and/or are not testing locally.

mia-ko replied:
Thanks Benjamin :)
Yeah, I know how they render. Personally I prefer code block rendition over the one using math, it's just something stylistically off with that one for me, for some reason I cannot quite explain. That's why I've been holding back.
But, I'll give it a go and see if it starts growing on me LOL

kartben commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,238 @@
+
+The lower transport layer takes Upper Transport PDUs from the upper transport layer and sends them to the lower transport layer on a peer device.
+If the Upper Transport PDU fits, it is sent in a single Lower Transport PDU.
+For longer packets, which do not fit into a single Lower Transport PDU, the lower transport layer performs segmentation, splitting the Upper Trasport
```
```suggestion
For longer packets, which do not fit into a single Lower Transport PDU, the lower transport layer performs segmentation, splitting the Upper Transport
```

Balaklaka commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,238 @@
+
+Four counters are related to SAR behavior:
+
+* Unicast retransmissions count (see `SAR Unicast Retransmissions Count`_ state and `SAR Unicast Retransmissions Without Progress Count`_ state)
```
```suggestion
* Two Unicast retransmissions counts (see `SAR Unicast Retransmissions Count`_ state and `SAR Unicast Retransmissions Without Progress Count`_ state)
```

Balaklaka commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,243 @@
+===================
+
+The value of this state defines the time since the last successfully received segment before giving up the reception of a segmented message.
+Use the Kconfig option :kconfig:option:`CONFIG_BT_MESH_SAR_RX_DISCARD_TIMEOUT` to set the value for this state.
```
What unit is this given in? (seconds/milliseconds)

mia-ko replied:
seconds

akredalen commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,243 @@
+* :kconfig:option:`CONFIG_BT_MESH_RX_SEG_MAX` to set the maximum number of segments in an incoming message.
+* :kconfig:option:`CONFIG_BT_MESH_TX_SEG_MAX` to set the maximum number of segments in an outcoming message.
+
+Never more than one segmented message is transmitted at the time, but the Bluetooth mesh stack allows you to define buffers that will queue a defined
```
Maybe something like this instead: "Only one segmented message is transmitted at a time, but the Bluetooth mesh stack allows you to **configure** buffers that can queue a specified number of incoming or outgoing segmented messages."

akredalen commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,243 @@
+of queued outgoing and incoming segmented messages respectively.
+
+Incoming and outgoing segmented messages share the same segment buffer, which means that they allocate their segments from the same pool.
+Both incoming and outgoing messages allocate segments at the start of the transaction, but release them at the different time.
```
I am a bit confused by this sentence. "but release them at the different time", what is meant by this? Should it be "at different times"?

mia-ko replied:
at different times, for sure :)

akredalen commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,243 @@
+******************************
+
+Transmission of segments is separated by a segment transmission interval (see the `SAR Segment Interval Step`_ state).
+Other time intervals and delays available for the segmentation and reassembly are:
```
```suggestion
Other configurable time intervals and delays available for the segmentation and reassembly are:
```

akredalen commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,243 @@
+
+The initial value of the SAR Multicast Retransmissions timer is set to the multicast retransmissions interval.
+
+When the lower transport layer receives a segment of a segmented message, it starts an SAR Discard timer.
```
```suggestion
When the lower transport layer receives a message segment, it starts a SAR Discard timer.
```

mia-ko replied:
It depends on whether you pronounce SAR as [sɑr] or [es/eɪ/ar/]. Not sure which one should be used.

akredalen replied:
I believe [sar] is the most common but this is just nitpicking :)

akredalen commented on doc/connectivity/bluetooth/api/mesh/sar_cfg_cli.rst:
```diff
@@ -0,0 +1,34 @@
+An element can send any SAR Configuration Client message at any time to query or change the states supported by the SAR Configuration Server model of a peer node.
+The SAR Configuration Client model only accepts messages encrypted with the device key of the node supporting the SAR Configuration Server model.
+
+If present, the SAR Configuration Client model must only be instantiated on the primary element.
```
I would like to keep the "only" here. So "shall only" then?

mia-ko replied:
I have checked docs for our other foundation models and I see that we've used
- must be 
- should be
- and can be
Most of them using must.
I wouldn't like to introduce yet another way of describing it. Are we ok with continuing with must only?

akredalen commented on doc/connectivity/bluetooth/api/mesh/sar_cfg_srv.rst:
```diff
@@ -0,0 +1,29 @@
+The SAR Configuration Server model does not have an API of its own, but relies on a :ref:`bluetooth_mesh_sar_cfg_cli` to control it.
+The SAR Configuration Server model only accepts messages encrypted with the node’s device key.
+
+If present, the model must be instantiated on the primary element.
```
See the comment above for the client.

mia-ko commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,238 @@
+Never more than one segmented message is transmitted at the time, but the Bluetooth mesh stack allows you to define buffers that will queue a defined
+number of ingoing or outgoing segmented messages.
+Use Kconfig options :kconfig:option:`CONFIG_BT_MESH_TX_SEG_MSG_COUNT` and :kconfig:option:`CONFIG_BT_MESH_RX_SEG_MSG_COUNT` to set the maximum number
+of queued outgoing and incoming segmented messages respectively.
```
:kconfig:option:`CONFIG_BT_MESH_TX_SEG_MSG_COUNT` defines outgoing messages, right? And that one is mentioned first in the sentence. That's why outgoing first, and then incoming. Would you like me to swap the Kconfig options to make your suggestion right?

mia-ko commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,238 @@
+* :kconfig:option:`CONFIG_BT_MESH_RX_SEG_MAX` to set the maximum number of segments in an incoming message.
+* :kconfig:option:`CONFIG_BT_MESH_TX_SEG_MAX` to set the maximum number of segments in an outcoming message.
+
+Never more than one segmented message is transmitted at the time, but the Bluetooth mesh stack allows you to define buffers that will queue a defined
+number of ingoing or outgoing segmented messages.
```
Right...
It's been a few ingoing and outcoming messages LOL Unbelievable!

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,247 @@
+
+(CONFIG_BT_MESH_SAR_RX_SEG_INT_STEP + 1) * 10 ms
+
+API reference
```
This can be removed. This API is not public.

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/sar_cfg_cli.rst:
```diff
@@ -0,0 +1,34 @@
+The SAR Configuration Client model is a foundation model defined by the Bluetooth mesh specification.
+It is an optional model, enabled with the :kconfig:option:`CONFIG_BT_MESH_SAR_CFG_CLI` configuration option.
+
+The SAR Configuration Client model is introduced in the `Bluetooth Mesh Protocol Specification version 1.1 <https://www.bluetooth.com/specifications/mesh-specifications/>`_,
```
Same comment as above. I'm not sure the link is needed in every file.

mia-ko replied:
I'm thinking having the link once, the first time it is introduced on a page, cannot hurt.
At least if looking from the perspective of our new doc portal, where the user will search for specific topics rather than read through page by page in a traditional manner. Having the link available on the page that user reads, instead of having to go back to a page where the link was introduced is better. We are incouraged to have an "every page is a first page" philosophy.
But I can remove it if you feel strongly about it.
Remove or stay?

PavelVPV replied:
What I **really** don't like is different styles in the same code base or documentation. What I expect will happen next is that here we will keep the link, but other pages will stay like they are right now (without the link). If we want to add link here, then we have to be consistent and add it to other pages as well (I don't see anything special in SAR models). 

I raised the concern of not having a common pattern/style for upstream docs on the last retrospective and created a follow up task NCSDK-23389. This particular case could be one of the points in that task.

P.S. This link doesn't really exist anymore, it redirects to all Bluetooth specs page.

mia-ko commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,243 @@
+SAR does not impose extra overhead on the access layer payload per segment.
+
+Segmentation and reassembly (SAR) Configuration models are foundation models defined by the
+`Bluetooth Mesh Protocol Specification version 1.1 <https://www.bluetooth.com/specifications/mesh-specifications/>`_.
```
This is the first and only place the spec is referenced on this page.
I am not sure what page you mean, maybe mesh.rst?
I also see we are only mentioning v1.0.1 there, no 1.1.

PavelVPV replied:
> This is the first and only place the spec is referenced on this page.
I am not sure what page you mean, maybe mesh.rst?

Yes, I meant mesh.rst. Otherwise, we need to add the reference to every model's doc page which is too much I think. 

> I also see we are only mentioning v1.0.1 there, no 1.1.

This should be fixed (not necessarily in your PR, but up to you).

mia-ko committed:
doc: bluetooth: mesh: listed models alphabetically

Changed placement of a few models so they are
in alphabetical order

Signed-off-by: Mia Koen <mia.koen@nordicsemi.no>

Balaklaka commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,276 @@
+address, but the transmission always succeedes after retransmitting all segments the configured
+number of times.
+
+The timeout between each retransmission to a unicast address and a group or virtual address is
+configured by Kconfig options :kconfig:option:`CONFIG_BT_MESH_TX_SEG_RETRANS_TIMEOUT_UNICAST` and
+:kconfig:option:`CONFIG_BT_MESH_TX_SEG_RETRANS_TIMEOUT_GROUP` respectively.
```
This was a little bit hard to follow, "unicast address and a group or virtual address". First there is "and" then "or"? Which kconfig option is used for virtual address?

akredalen commented on doc/connectivity/bluetooth/api/mesh/sar_cfg_srv.rst:
```diff
@@ -0,0 +1,29 @@
+The SAR Configuration Server model does not have an API of its own, but relies on a :ref:`bluetooth_mesh_sar_cfg_cli` to control it.
+The SAR Configuration Server model only accepts messages encrypted with the node’s device key.
+
+If present, the model must only be instantiated on the primary element.
```
I suggested adding the word "only" to emphasize that the model cannot be instantiated on any other element. Without the word "only," it might be interpreted as allowing you to also adding it to the next elements as long as the model is present on the primary element.

The spec says: "If supported, the model **shall** be supported by a primary element **and** shall not be supported by any secondary elements."

Balaklaka reviewed (approved):
I am approving this, as my only comment is about this "must vs shall" and "only" wording.

mia-ko committed:
doc: bluetooth: mesh: fixed spec links

Corrected the profile specification v1.0.1 link to point to
the spec directly and not to the general specification page.
Added link to protocol specification v1.1.

Signed-off-by: Mia Koen <mia.koen@nordicsemi.no>

PavelVPV commented on doc/connectivity/bluetooth/api/mesh.rst:
```diff
@@ -5,7 +5,10 @@ Bluetooth Mesh Profile
-`Bluetooth Mesh Profile Specification v1.0.1 <https://www.bluetooth.com/specifications/mesh-specifications/>`_.
+`Bluetooth Mesh Profile Specification v1.0.1 <https://www.bluetooth.com/specifications/specs/mesh-profile-1-0-1/>`_.
+
+Implementation of the draft `Bluetooth Mesh Protocol Specification v1.1 <https://www.bluetooth.com/specifications/specs/mesh-protocol-1-1/>`_
```
Something with rendering here

PavelVPV replied:
Sorry, now it looks fine. In one of the previous builds it didn't render the link correctly.

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/sar_cfg.rst:
```diff
@@ -0,0 +1,278 @@
+
+.. math::
+
+   [unicast~retransmissions~interval~step + unicast~retransmissions~interval~increment \times (TTL - 1)]
```
Minor: do we need square brackets here?

PavelVPV replied:
Yes

mia-ko commented:
> I am approving this, as my only comment is about this "must vs shall" and "only" wording.

I think we can keep 'must'. It is not wrong, and within Nordic we have this rule of avoiding shall/should and using must/will instead. And even though this is not NCS documentation, I am still quite influenced by this rule.
When it comes to 'only', I agree with Stine on this one. Stating that something must exist in one place, does not explicitly mean that it cannot exist elsewhere in addition. Using only, we make sure that it is on primary only. If that is ok with you, we do no changes.

PavelVPV commented:
> > I am approving this, as my only comment is about this "must vs shall" and "only" wording.
> 
> I think we can keep 'must'. It is not wrong, and within Nordic we have this rule of avoiding shall/should and using must/will instead. And even though this is not NCS documentation, I am still quite influenced by this rule. When it comes to 'only', I agree with Stine on this one. Stating that something must exist in one place, does not explicitly mean that it cannot exist elsewhere in addition. Using only, we make sure that it is on primary only. If that is ok with you, we do no changes.

We need to align other models to this rule as some don't have "only", some has "can be" instead of "must"...

PavelVPV commented:
> > > I am approving this, as my only comment is about this "must vs shall" and "only" wording.
> > 
> > 
> > I think we can keep 'must'. It is not wrong, and within Nordic we have this rule of avoiding shall/should and using must/will instead. And even though this is not NCS documentation, I am still quite influenced by this rule. When it comes to 'only', I agree with Stine on this one. Stating that something must exist in one place, does not explicitly mean that it cannot exist elsewhere in addition. Using only, we make sure that it is on primary only. If that is ok with you, we do no changes.
> 
> We need to align other models to this rule as some don't have "only", some has "can be" instead of "must"...

I'll fix this separately.

akredalen commented:
> > We need to align other models to this rule as some don't have "only", some has "can be" instead of "must"...
> 
> I'll fix this separately.

I think @mia-ko is on this already and wanted to add an additional commit to this PR?

PavelVPV commented:
> > > We need to align other models to this rule as some don't have "only", some has "can be" instead of "must"...
> > 
> > 
> > I'll fix this separately.
> 
> I think @mia-ko is on this already and wanted to add an additional commit to this PR?

Sorry, @mia-ko , I created it already: https://github.com/zephyrproject-rtos/zephyr/pull/62523

mia-ko commented:
> > > > We need to align other models to this rule as some don't have "only", some has "can be" instead of "must"...
> > > 
> > > 
> > > I'll fix this separately.
> > 
> > 
> > I think @mia-ko is on this already and wanted to add an additional commit to this PR?
> 
> Sorry, @mia-ko , I created it already: #62523

It's ok, as long as it gets done :) Thanks. I'll revert my changes.
